internal/formats/common/cyclonedxhelpers/component.go

```diff
@@ -1,27 +1,164 @@
 package cyclonedxhelpers
 
 import (
+	"fmt"
+	"reflect"
+	"strconv"
+
 	"github.com/CycloneDX/cyclonedx-go"
+
+	"github.com/anchore/syft/internal/log"
 	"github.com/anchore/syft/syft/pkg"
+	"github.com/anchore/syft/syft/source"
 )
 
-func Component(p pkg.Package) cyclonedx.Component {
+func encodeComponent(p pkg.Package) cyclonedx.Component {
 	return cyclonedx.Component{
 		Type:               cyclonedx.ComponentTypeLibrary,
 		Name:               p.Name,
-		Group:              Group(p),
+		Group:              encodeGroup(p),
 		Version:            p.Version,
 		PackageURL:         p.PURL,
-		Licenses:           Licenses(p),
-		CPE:                CPE(p),
-		Author:             Author(p),
-		Publisher:          Publisher(p),
-		Description:        Description(p),
-		ExternalReferences: ExternalReferences(p),
-		Properties:         Properties(p),
+		Licenses:           encodeLicenses(p),
+		CPE:                encodeCPE(p),
+		Author:             encodeAuthor(p),
+		Publisher:          encodePublisher(p),
+		Description:        encodeDescription(p),
+		ExternalReferences: encodeExternalReferences(p),
+		Properties:         encodeProperties(p),
 	}
 }
 
 func hasMetadata(p pkg.Package) bool {
 	return p.Metadata != nil
 }
+
+func decodeComponent(c *cyclonedx.Component) *pkg.Package {
+	typ := pkg.Type(findPropertyValue(c, "type"))
+	purl := c.PackageURL
+	if typ == "" && purl != "" {
+		typ = pkg.TypeFromPURL(purl)
+	}
+
+	metaType, meta := decodePackageMetadata(c)
+
+	p := &pkg.Package{
+		Name:         c.Name,
+		Version:      c.Version,
+		FoundBy:      findPropertyValue(c, "foundBy"),
+		Locations:    decodeLocations(c),
+		Licenses:     decodeLicenses(c),
+		Language:     pkg.Language(findPropertyValue(c, "language")),
+		Type:         typ,
+		CPEs:         decodeCPEs(c),
+		PURL:         purl,
+		MetadataType: metaType,
+		Metadata:     meta,
+	}
+
+	return p
+}
+
+func decodeLocations(c *cyclonedx.Component) (out []source.Location) {
+	if c.Properties != nil {
+		props := *c.Properties
+		for i := 0; i < len(props)-1; i++ {
+			if props[i].Name == "path" && props[i+1].Name == "layerID" {
```

Comment:
This is a pretty janky thing: we have multiple props with the same name -- each location struct is output with path=.. layerID=..; we should figure out if there is a better way to accomplish this

Reply:
I have been thinking about this from when I implemented this. What if we added the index in the property name, so path.0, layerId.0 and so on for any array like elements? Or maybe we can even utilize the syntax from gron https://github.com/tomnomnom/gron#usage ?